src/sentry/utils/snowflake.py

```diff
@@ -55,3 +94,55 @@ def snowflake_id_generation():
     ID_VALIDATOR.validate(snowflake_id)
 
     return snowflake_id
+
+
+def get_redis_cluster(redis_key: str):
+    return redis.clusters.get("default").get_local_client_for_key(redis_key)
+
+
+def store_snowflake_to_redis(redis_key: str, timestamp: int, used_region_sequence: int) -> None:
+    cluster = get_redis_cluster(redis_key)
+    used_region_sequences = cluster.get(str(timestamp))
+    if used_region_sequences:
+        used_region_sequences = json.loads(used_region_sequences)
+    else:
+        used_region_sequences = []
+    used_region_sequences.append(used_region_sequence)
+    cluster.setex(str(timestamp), int(_TTL.total_seconds()), json.dumps(used_region_sequences))
+
+
+def get_snowflake_from_redis(redis_key: str, starting_timestamp: int) -> Tuple[int, int]:
+    cluster = get_redis_cluster(redis_key)
+
+    for i in range(int(_TTL.total_seconds())):
+        timestamp = starting_timestamp - i
+
+        used_region_sequences = cluster.get(str(timestamp))
+
+        if not used_region_sequences:
+            used_region_sequences = []
+        else:
+            used_region_sequences = json.loads(used_region_sequences)
+
+        if len(used_region_sequences) == (MAX_AVAILABLE_REGION_SEQUENCEs):
+            continue
+
+        for region_sequence in get_region_sequence_assignment_order():
+            if region_sequence not in used_region_sequences:
+                store_snowflake_to_redis(redis_key, timestamp, region_sequence)
+                return timestamp, region_sequence
```

Comment:
This is where we're intentionally tolerating a race condition, so we should document it.

For example, the following might happen:

    Redis contains:   [9, 13]
    Server 1 reads:   [9, 13]
    Server 2 reads:   [9, 13]
    Server 1 assigns: 15
    Server 2 assigns: 6
    Server 1 writes:  [9, 13, 15]
    Server 2 writes:  [9, 13, 6]
    Redis contains:   [9, 13, 6]

Now 15 is assigned, but Redis fails to reflect it. This is fine, as long as Postgres is the final source of truth and we can gracefully retry if a collision happens on write. Redis is still useful as long as it *usually* warns us of a collision before calling out to Postgres.

The docstring or comments in this function should indicate:
1. There is a race condition between the Redis read (line 120) and write (line 132), which we are aware of and accept as a "good enough" solution.
2. The return value from this function only indicates that the region sequence value is *probably* not already taken, and should be used only to improve efficiency, not as a source of truth.

Reply:
I don't know if we need to track the consumed values and parse/serialize JSON. Wouldn't a counter that is atomically incremented handle what we need?

Handling the upper bound of the region sequence would still be required, but that is also true for this implementation.

Reply:
If we're using redis, why are we using json like this? Can't we just push/pop from a redis list?

Reply:
👍 I wasn't fully aware if Redis's features would support that. Having Redis increment the number atomically is the way to go.